src/singularity_runners/cli.py

```diff
@@ -0,0 +1,37 @@
+import argparse
+import sys
+
+from . import github, gitlab
+
+
+def subparser(parser: argparse.ArgumentParser):
+    subparser = parser.add_subparsers(dest='provider', required=True)
+
+    github.subparser(subparser.add_parser('github'))
+    gitlab.subparser(subparser.add_parser('gitlab'))
+
+    return parser
+
+
+def main():
+    parser = subparser(argparse.ArgumentParser(add_help=False))
+
+    args, _ = parser.parse_known_args()
+
+    selected_parser = parser._actions[0].choices[args.provider]
```

Comment:
`_actions` is technically private (by convention, names prefixed with an underscore are considered private). Argparse is probably not changing very much, so you might get away with it, but this - and the overall complexity of the argument parsing here - feels like a code smell.

Given that this runs inside the container, so the main command is encoded in the container recipe rather than something you type frequently, I'd get rid of at least one level of subcommands (i.e. `singularity-gh-runner start`) to simplify the implementation.

I think argparse is designed with the idea that you specify all of your CLI up front, not lazily parsing piece by piece in different modules. I also [experimented with](https://github.com/takluyver/flit/pull/161/files) splitting up subcommand definitions for Flit some time ago, but then decided it wasn't worth the trouble (which is why that PR isn't merged).

Reply:
Yeah I think I'll end up getting rid of a lot of the argparse complexity, it only ended up happening since I thought doing it like that would be neat and then stubbornly kept going down that route even once I saw that it was more complex than I expected, since it *seemed* like it shouldn't be too complex or difficult...

Reply:
I'm still sure that there must be a better way to do this since to me this kind of usecase seems pretty standard but I guess maybe it isn't.

Reply:
I don't think it's meant to be necessary to get the selected subparser - the examples in the docs show just using the result from the top-level parser, which includes options from subparsers. Possibly the confusion is trying to use the same `main()` function both for handing off subcommands after parsing command line args, and when running e.g. `python -m singularity_runners.github.cli` directly?